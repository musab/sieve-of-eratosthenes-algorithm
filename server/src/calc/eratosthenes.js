// https://stackoverflow.com/a/15471749/3330490
// Eratosthenes algorithm to find all primes under n

export const eratosthenes = n => {
  if (n < 1 || !Number(n)) {
    throw new Error('Must supply a Number that is greater than 1');
  }

  const array = [];
  const output = [];
  const upperLimit = Math.sqrt(n);

  // Make an array from 2 to (n - 1)
  for (var i = 0; i < n; i++) {
    array.push(true);
  }

  // Remove multiples of primes starting from 2, 3, 5,...
  for (var i = 2; i <= upperLimit; i++) {
    if (array[i]) {
      for (var j = i * i; j < n; j += i) {
        array[j] = false;
      }
    }
  }

  // All array[i] set to true are primes
  for (var i = 2; i < n; i++) {
    if (array[i]) {
      output.push(i);
    }
  }

  return output;
};
