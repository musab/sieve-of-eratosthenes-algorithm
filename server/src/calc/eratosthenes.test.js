import { eratosthenes } from './eratosthenes';

describe('eratosthenes', () => {
  test('10 returns [ 2, 3, 5, 7 ]', () => {
    const input = 10;
    const expected = [2, 3, 5, 7];

    expect(eratosthenes(input)).toStrictEqual(expected);
  });

  test('18 returns [2, 3, 5, 7, 11, 13, 17]', () => {
    const input = 18;
    const expected = [2, 3, 5, 7, 11, 13, 17];

    expect(eratosthenes(input)).toStrictEqual(expected);
  });

  test('1 returns []', () => {
    const input = 1;
    const expected = [];

    expect(eratosthenes(input)).toStrictEqual(expected);
  });

  test('2 returns []', () => {
    const input = 2;
    const expected = [];

    expect(eratosthenes(input)).toStrictEqual(expected);
  });

  test('throws if called unparsable Number ', () => {
    const input = 'x1x';
    function passString() {
      eratosthenes(input);
    }

    expect(passString).toThrowErrorMatchingSnapshot();
  });

  test('throws if called negative Number', () => {
    const input = -5;
    function passNegativeNum() {
      eratosthenes(input);
    }

    expect(passNegativeNum).toThrowErrorMatchingSnapshot();
  });
});
