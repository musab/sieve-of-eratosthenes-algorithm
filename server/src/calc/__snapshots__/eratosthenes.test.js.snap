// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`eratosthenes throws if called negative Number 1`] = `"Must supply a Number that is greater than 1"`;

exports[`eratosthenes throws if called with NaN 1`] = `"Must supply a Number that is greater than 1"`;
